File: 4/4.dart
bool check(num) {
  String s = num.toString();
  bool c = false;
  for (int i = 1; i < s.length; i++) {
    if (s[i - 1] == s[i]) {
      c = true;
    }
    if (int.parse(s[i - 1]) > int.parse(s[i])) {
      return false;
    }
  }
  return c;
}

RegExp dupeRegex = new RegExp('(\\d)\\1+');
bool check2(num) {
  String s = num.toString();
  bool c = false;
  for (int i = 1; i < s.length; i++) {
    for (var match in dupeRegex.allMatches(s)) {
      if (match.group(0).length == 2) {
        c = true;
      }
    }
    if (int.parse(s[i - 1]) > int.parse(s[i])) {
      return false;
    }
  }
  return c;
}

void main() {
  int tot = 0;
  for (int i = 307237; i <= 769058; i++) {
    if (check(i)) {
      tot += 1;
    }
  }
  print('Answer to part 1: $tot');
  tot = 0;
  for (int i = 307237; i <= 769058; i++) {
    if (check2(i)) {
      tot += 1;
    }
  }
  print('Answer to part 2: $tot');
}

File: 4/4.py
import sys, re

def check(num):
    s = str(num)
    c = False
    for i in range(1, len(s)):
        if s[i-1] == s[i]:
            c = True
        if int(s[i-1]) > int(s[i]):
            return False
    return c

def check2(num):
    s = str(num)
    c = False
    last = ''
    dupes = re.finditer(r'(\d)\1+', s)
    for i in dupes:
        for j in range(len(i.groups())):
            if len(i.group(j)) == 2:
                c = True
    if not c: return False
    for i in range(1, len(s)):
        last = s[i-1]
        if int(s[i-1]) > int(s[i]):
            return False
    return c


tot = 0
for i in range(307237, 769058):
    if check(i):
        tot += 1
print("Answer to part 1: {}".format(tot))

tot = 0
for i in range(307237, 769058):
    if check2(i):
        tot += 1
print("Answer to part 2: {}".format(tot))


